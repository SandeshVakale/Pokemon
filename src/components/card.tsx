import React, {useEffect, useState} from 'react';
import {Dimensions, Image, Text, StyleSheet, View} from 'react-native';
import {result} from '../store/list';
import {LinearGradient} from 'expo-linear-gradient';
import {ColorModifier} from '../utils/colorModifier';
import {typography} from '../theme';
import {GetSpeciesDetails} from '../utils/getSpecies';

const CardMemo = ({index, pokemon}: {index: number; pokemon: result}) => {
  const {h6, h7} = typography;

  const [species, setSpecies] = useState<any>(null);
  useEffect(() => {
    GetSpeciesDetails(pokemon.name)?.then(data => setSpecies(data.data));
  });

  const {lightColor, darkColor, mainColor} = ColorModifier(
    species?.color?.name,
  );
  return (
    <LinearGradient
      colors={species ? [mainColor, darkColor] : ['yellow', 'gold']}
      key={index}
      style={styles.container}>
      <Image source={{uri: pokemon.url}} style={styles.image} />
      <View
        style={[
          styles.circle,
          {
            backgroundColor: lightColor,
          },
        ]}
      />
      <Text style={[h6, styles.text]}>{pokemon.name.toLocaleUpperCase()}</Text>
      {species?.shape && (
        <Text style={[h7, styles.textSubtitle]}>{species.shape.name}</Text>
      )}
    </LinearGradient>
  );
};

function arePropsEqual(
  prevProps: {pokemon: result},
  nextProps: {pokemon: result},
) {
  /*
    return true if passing nextProps to render would return
    the same result as passing prevProps to render,
    otherwise return false
    */
  return (
    nextProps.pokemon.name === prevProps.pokemon.name &&
    nextProps.pokemon.url === prevProps.pokemon.url
  );
}

export const Card = React.memo(CardMemo, arePropsEqual);

const styles = StyleSheet.create({
  container: {
    height: Dimensions.get('window').height * 0.2,
    width: Dimensions.get('window').width * 0.45,
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
    borderRadius: 20,
    padding: 15,
    margin: 10,
    overflow: 'hidden',
  },
  image: {
    position: 'absolute',
    right: -20,
    bottom: 0,
    zIndex: 100,
    height: Dimensions.get('window').height * 0.12,
    width: Dimensions.get('window').width * 0.4,
    resizeMode: 'contain',
  },
  text: {
    color: 'white',
  },
  textSubtitle: {
    color: 'white',
    padding: 5,
  },
  circle: {
    height: 120,
    width: 120,
    borderRadius: 60,
    position: 'absolute',
    right: -20,
    bottom: -20,
  },
});
