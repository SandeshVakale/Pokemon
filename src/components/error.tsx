import React from 'react';
import {SafeAreaView, StyleSheet, Text} from 'react-native';
import {typography, colors} from '../theme';
import {MaterialIcons} from '@expo/vector-icons';
import {errorType} from '../types';

export const Error = ({error}: {error: errorType}) => {
  const {h3, h6} = typography;
  const {red} = colors;
  return (
    <SafeAreaView style={styles.container}>
      <MaterialIcons name="error" size={60} color={red} />
      <Text style={h3}>{error.code}</Text>
      <Text style={h6}>{error.message}</Text>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.white,
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export default Error;
