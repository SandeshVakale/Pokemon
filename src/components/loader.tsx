import React from 'react';
import {SafeAreaView, ActivityIndicator, StyleSheet} from 'react-native';
import {colors} from '../theme';

export const Loader = () => {
  return (
    <SafeAreaView style={styles.container}>
      <ActivityIndicator />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.white,
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export default Loader;
