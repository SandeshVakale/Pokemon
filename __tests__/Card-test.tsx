import 'react-native';
import React from 'react';
import {Card} from '../src/components/card';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('renders correctly', async () => {
  const tree = renderer
    .create(
      <Card
        index={1}
        pokemon={{
          name: 'Sandesh',
          url: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png',
        }}
        isColored={false}
      />,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
