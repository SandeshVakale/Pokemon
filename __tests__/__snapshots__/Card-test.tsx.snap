// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<ViewManagerAdapter_ExpoLinearGradient
  proxiedProperties={
    Object {
      "colors": Array [
        4294967040,
        4294956800,
      ],
      "endPoint": undefined,
      "locations": undefined,
      "startPoint": undefined,
    }
  }
  style={
    Object {
      "alignItems": "flex-start",
      "borderRadius": 20,
      "height": 266.8,
      "justifyContent": "flex-start",
      "margin": 10,
      "overflow": "hidden",
      "padding": 15,
      "width": 337.5,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png",
      }
    }
    style={
      Object {
        "bottom": 0,
        "height": 160.07999999999998,
        "position": "absolute",
        "resizeMode": "contain",
        "right": -20,
        "width": 300,
        "zIndex": 100,
      }
    }
  />
  <View
    style={
      Array [
        Object {
          "borderRadius": 60,
          "bottom": -20,
          "height": 120,
          "position": "absolute",
          "right": -20,
          "width": 120,
        },
        Object {
          "backgroundColor": "yellow",
        },
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Lexend_400Regular",
          "fontSize": 20,
        },
        Object {
          "color": "white",
        },
      ]
    }
  >
    SANDESH
  </Text>
</ViewManagerAdapter_ExpoLinearGradient>
`;
